text_classification.py

```diff
@@ -0,0 +1,487 @@
+    def __len__(self):
+        try:
+            return len(self.examples)
+        except TypeError:
+            return 2**32
+
+    def __iter__(self):
+        for x in self.examples:
+            yield x
+
+    def splits(self):
+        """ Create dataset objects for splits of the dataset.
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def sort_key(ex):
+        return len(ex.text)
+
+    def split(self, split_ratio=0.7, random_state=None):
+        """ Create train-test-valid splits from the instance's examples.
+            The initial dataset has no valid set. Therefore, the train set
+            is split into train and valid parts.
+
+        Inputs:
+            split_ratio: the percent out of the initial train set split into
+                train dataset objects. Default: 0.7
+
+        Examples:
+            >>> train_data, test_data, valid_data = txt_cls.split()
+
+        Outputs:
+            - train_data: a datset objects based on
+                train.csv file with split_ratio percent.
+            - train_data: a datset objects based on test.csv file.
+            - valid_data: a datset objects based on
+                train.csv file with 1-split_ratio percent.
+
+        """
+        rnd = dataset.RandomShuffler(random_state)
+        train_ratio, test_ratio, valid_ratio = split_ratio, 0.0, 1 - split_ratio
+        train_examples, test_examples, valid_examples = \
+            dataset.rationed_split(self.train_examples,
+                                   train_ratio, test_ratio, valid_ratio, rnd)
+        test_examples = self.test_examples
+        split = tuple(TextDataset(d, self.fields)
+                      for d in (train_examples, test_examples, valid_examples)
+                      if d is not None)
+        for subset in split:
+            subset.sort_key = self.sort_key
+
+        return split
+
+    def iters(self, split_ratio=0.7, batch_size=32, device='cpu', random_state=None):
+        """Create iterator objects for splits of the dataset.
+
+        Arguments:
+            split_ratio: split train_examples into train set (split_ratio)
+                and valid set (1-split_ratio). Default: 0.7
+            batch_size: batch size. Default: 32
+            device: the device to sent data. Default: 'cpu'
+            random_state: the random state provided by user. Default: None
+
+        Examples:
+            >>> train_iter, test_iter, valid_iter = txt_cls.iters(device="cpu")
+
+        Outputs:
+            - train_iter: a iterator based on
+                train.csv file with split_ratio percent.
+            - train_iter: a iterator based on test.csv file.
+            - valid_iter: a iterator based on
+                train.csv file with 1-split_ratio percent.
+
+        """
+
+        train, test, valid = self.split(split_ratio, random_state)
+        return generate_iterators(
+            (train, test, valid), [batch_size] * 3, device=device)
+
+    def download(self):
```

Comment:
Can we move this out of TextClassificationDataset?

Reply:
yeap, download function is now independent and can stay outside.